editor/hooks/autocomplete/index.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * External dependencies
+ */
+import { compose } from 'lodash';
+import { connect } from 'react-redux';
+
+/**
+ * WordPress dependencies
+ */
+import { Autocomplete, withContext } from '@wordpress/components';
+import { addFilter } from '@wordpress/hooks';
+
+/**
+ * Internal dependencies
+ */
+import { blockAutocompleter, userAutocompleter } from './autocompleters';
+import { getInserterItems } from '../../store/selectors';
+
+function withAutocomplete( BlockEdit ) {
+	function WrappedBlockEdit( props ) {
+		const { name, items, onReplace } = props;
+
+		// Only wrap paragraph blocks with <Autocomplete>
+		if ( name !== 'core/paragraph' ) {
```

Comment:
It makes a lot of sense for `blockAutocompleter` to be applied to a paragraph block only, but not sure if we want to keep this limitation for `userAutocompleter`, too.

We could make it more customizable by using [supports proprty](https://github.com/WordPress/gutenberg/blob/master/docs/block-api.md#supports-optional) from the block definition. One way of doing it would be:

```js
name: 'core/paragraph',
supports: {
   autocompleters: [ 'block', 'user' ]
},
```
```js
name: 'core/list',
supports: {
   autocompleters: [ 'user' ]
},
```

Reply:
This is an interesting idea, @gziolo. It's better to declare autocompleters explicitly than to bury them in block implementation.

Would it be better to provide the actual completers instead of strings? Do we gain anything from the indirection of using strings?

If we specify an actual list of completers, we could override aspects of them using the `blocks.registerBlockType` filter rather than the dedicated filter discussed by #4609. The initial filters I proposed there were too granular for a first pass, but maybe using `blocks.registerBlockType` for this isn't granular enough. :)

Reply:
> Would it be better to provide the actual completers instead of strings? Do we gain anything from the indirection of using strings?

Yes, this is also an option to define them as part of the block API. I like how this discussion goes. Very good ideas :)

We do something similar with `transformers`. See here: https://github.com/WordPress/gutenberg/blob/727da2c415f1768f7aa681fff6f9accbd9177d00/blocks/library/code/index.js#L37. So might be a way to go in this case, too. You wouldn't have to create new hooks, you could use  `blocks.registerBlockType` as described [here](https://github.com/WordPress/gutenberg/blob/master/docs/extensibility.md#modifying-blocks-experimental).

Reply:
/cc @aduth @youknowriad

Reply:
Having to add autocompleters more broadly was one of the initial reasons for the refactoring of these autocompleters. so yes 💯 

I like the`supports`API thought as it can be easily moved to the backend for instance and I see other blocks using the "block" autocompleter as well (for example heading).

Reply:
If I comment out this check then it seems to work OK with blocks that have multiple `RichText` components:

![autocomplete](https://user-images.githubusercontent.com/612155/35785103-85556bca-0a69-11e8-815d-76f776c5571a.gif)

And it doesn't seem to affect blocks that use `<input>`s and `<textarea`>s:

![autocomplete-2](https://user-images.githubusercontent.com/612155/35785185-4ddddbb8-0a6a-11e8-9861-ede2d43ec2de.gif)

So I think it would be reasonable to to have all blocks wrapped with the user autocompleter and only paragraph blocks wrapped with the block autocompleter.